### PR TITLE
feat: adjust theme types to match internal and regular usage

### PR DESCRIPTION
**Pull request by Drakeoon**
During the development we came to the conclusion that `useTheme` right now has to serve too many roles - support two themes at the same time for internal components and later on support just one theme when used by developers including `react-native-paper` into their applications

### Summary

- [x] Added `useInternalTheme` hook for internal usage
- [x] Added `withInternalTheme` HOC for internal usage
- [x] Renamed global type definition
- [x] Change usage inside internal components
- [x] Raised a [PR](https://github.com/callstack/react-theme-provider/pull/143) in `@callstack/react-theme-provider` to allow typing hooks
- [ ] TypeScript support for custom themes for withTheme HOC is still in progress, probably not a part of this PR

### Test plan

Due to the usage of global `index.d.ts` files I had troubles with testing the implementation inside `example` folder. My idea now is to create a separate app, include `react-native-paper` from this branch and see if global theme overrides work properly

### CODE REVIEW

**Comment by Drakeoon** [comment]
@lukewalczak CircleCI tests fail due to missing implementation in `@callstack/react-theme-provider` - I'll raise the PR today

File: docs/pages/2.theming.md [hunk 1]
```diff
@@ -178,26 +181,100 @@ export default function Main() {
 
 ## TypeScript
 
-By default extending the theme won't work well with TypeScript, but we can take advantage of `global augmentations` and specify the new properties that we added to the theme:
+By default TypeScript works well whenever you change the value of built-in theme's properties. It gets more complicated when you want to extend the theme's properties or change their types. In order to fully support TypeScript, you will need to follow the guide that fits your use-case most accurately:
```
Comment (lukewalczak) [inline docs/pages/2.theming.md:184]:
```suggestion
By default, TypeScript works well whenever you change the value of the built-in theme's properties. It gets more complicated when you want to extend the theme's properties or change their types. In order to fully support TypeScript, you will need to follow the guide that fits your use-case most accurately:
```

File: docs/pages/2.theming.md [hunk 2]
```diff
@@ -178,26 +181,100 @@ export default function Main() {
...
+
+There are two supported ways of overriding theme:
```
Comment (lukewalczak) [inline docs/pages/2.theming.md:186]:
```suggestion
There are two supported ways of overriding the theme:
```

File: docs/pages/2.theming.md [hunk 3]
```diff
@@ -178,26 +181,100 @@ export default function Main() {
...
+
+1. **Simple built-in theme overrides** - when you only customize the values and the whole theme schema remains the same
+2. **Advanced theme overrides** - when you _add new properties_ or _change the built-in schema shape_
+
+### Simple built-in theme overrides
+
+You can provide a `theme` prop with a theme object with same properties as the default theme:
```
Comment (lukewalczak) [inline docs/pages/2.theming.md:193]:
```suggestion
You can provide a `theme` prop with a theme object with the same properties as the default theme:
```

File: docs/pages/2.theming.md [hunk 4]
```diff
@@ -178,26 +181,100 @@ export default function Main() {
...
+    <PaperProvider theme={theme}>
+      <App />
+    </PaperProvider>
+  );
+}
+```
+
+### Advanced theme overrides
+
+If you need to modify the built-in theme schema by adding a new property or changing it type, you need to follow these steps:
```
Comment (lukewalczak) [inline docs/pages/2.theming.md:222]:
```suggestion
If you need to modify the built-in theme schema by adding a new property or changing its type, you need to follow these steps:
```

File: docs/pages/2.theming.md [hunk 5]
```diff
@@ -178,26 +181,100 @@ export default function Main() {
...
+  );
+}
+```
+
+2. Create a typed `useAppTheme()` hook in your project
 
+```ts
 import * as React from 'react';
-import { MD3LightTheme as DefaultTheme, Provider as PaperProvider } from 'react-native-paper';
+import { MD3LightTheme, Provider as PaperProvider } from 'react-native-paper';
```
Comment (lukewalczak) [inline docs/pages/2.theming.md:258]:
```suggestion
import { MD3LightTheme, Provider as PaperProvider, useTheme } from 'react-native-paper';
```

File: docs/pages/2.theming.md [hunk 6]
```diff
@@ -242,72 +324,102 @@ export default function Main() {
 
 Specifying `{ version: 2 }` tells React Native Paper to use the built in Material Design 2 theme, so you don't have to fully extend it on your own.
 
-### Typescript
+### Advanced setup
 
-You need to apply a global type change in a similar way how you would do that when [extending the theme](#typescript)
+As with any theme, you can also specify your custom properties within Material Design 2 theme:
```
Comment (lukewalczak) [inline docs/pages/2.theming.md:329]:
```suggestion
As with any theme, you can also specify your custom properties within the Material Design 2 theme:
```

File: docs/pages/2.theming.md [hunk 7]
```diff
@@ -208,23 +285,28 @@ export default function Main() {
...
 
 ## Material Design 2
 
-If you want to use previous versions of Material Design **we provide backwards support for older versions**.
+Using Material Design 2 is **fully supported in React Native Paper v5+**.
+
+### Simple setup
 
-In order to change the theme version, you can pass it as a prop to the `Provider` like so:
+In order to use Material Design 2 theme you can just pass `{ version: 2 }` to the PaperProvider theme prop:
```
Comment (lukewalczak) [inline docs/pages/2.theming.md:309]:
```suggestion
In order to use the Material Design 2 theme you can just pass `{ version: 2 }` to the PaperProvider theme prop:
```

File: docs/pages/2.theming.md [hunk 8]
```diff
@@ -242,72 +324,102 @@ export default function Main() {
...
+Due to the amount of changes in the theme's schema shape it falls into the [Advanced theme overrides](#advanced-theme-overrides) category. The steps are identical as with any advanced theme, just make sure to extend the built-in `MD2LightTheme` or `MD2DarkTheme` instead of `MD3LightTheme` or `MD3DarkTheme`.
 
-If you want to change the theme for a certain component from the library, you can directly pass the `theme` prop to the component. The theme passed as the prop is merged with the theme from the `Provider`:
+The final example for Material Design 2 would look like this:
 
-```js
+```ts
 import * as React from 'react';
-import { Button } from 'react-native-paper';
+import { MD2LightTheme, Provider as PaperProvider } from 'react-native-paper';
```
Comment (lukewalczak) [inline docs/pages/2.theming.md:366]:
```suggestion
import { MD2LightTheme, Provider as PaperProvider, useTheme } from 'react-native-paper';
```

File: docs/pages/2.theming.md [hunk 9]
```diff
@@ -208,23 +285,28 @@ export default function Main() {
 }
 ```
 
+3. Use the `useAppTheme()` hook in your components
```
Comment (lukewalczak) [inline docs/pages/2.theming.md:288]:
```suggestion
3. Start using the `useAppTheme()` hook across your components in the whole app
```

or something similar, just want to **highlight** that then `useAppTheme` should be the one and only hook for getting the theme

**Review by lukewalczak — COMMENTED** [review]
Small grammar suggestion and adding `useTheme` in example

**Review by lukewalczak — APPROVED** [review]
Great job @Drakeoon 🔥!

**Pull request merged**